GDBM_File: Implement crash-tolerance and export/import functions.

graygnuorg opened the pull request:
* ext/GDBM_File/Makefile.PL: Register new constants: gdbm_open
flags and return values for gdbm_latest_snapshot.
* ext/GDBM_File/GDBM_File.pm: Update documentation.
Export new constants.
Raise version to 1.21.
* ext/GDBM_File/GDBM_File.xs (dbcroak): Include system error
infomation, when appropriate.
(gdbm_syserrno): Return a meaningful value if not_here.
(gdbm_dump, gdbm_load, gdbm_convert)
(gdbm_failure_atomic, gdbm_latest_snapshot)
(gdbm_crash_tolerance_status): New functions.
* ext/GDBM_File/t/dump.t: New testcase.
* ext/GDBM_File/t/snapshot.t: New testcase.
* MANIFEST: List new files.

graygnuorg committed:
GDBM_File: Implement crash-tolerance and export/import functions.

* ext/GDBM_File/Makefile.PL: Register new constants: gdbm_open
flags and return values for gdbm_latest_snapshot.
* ext/GDBM_File/GDBM_File.pm: Update documentation.
Export new constants.
Raise version to 1.21.
* ext/GDBM_File/GDBM_File.xs (dbcroak): Include system error
infomation, when appropriate.
(gdbm_syserrno): Return a meaningful value if not_here.
(gdbm_dump, gdbm_load, gdbm_convert)
(gdbm_failure_atomic, gdbm_latest_snapshot)
(gdbm_crash_tolerance_status): New functions.
* ext/GDBM_File/t/dump.t: New testcase.
* ext/GDBM_File/t/snapshot.t: New testcase.
* MANIFEST: List new files.

graygnuorg commented:
Hi, Jim

I was able to build the stuff on FreeBSD 12.0 (and 13.0, for what it's worth) without any problem.  In your build case, I don't know what regular_configure is, so I can only suppose that there's something wrong with the configuration options it uses.  If you can give me some insight into what it does, perhaps I could investigate it further.

Re "recent years maintaining GDBM_File has been a PITA", my impression is that too little effort has been invested in maintaining it.  That's why I started to do it myself.  As a maintainer of GNU dbm I'm interested in that GDBM_File adequately reflects the gdbm functionality and provides the same possibilities.

As to the last question (why this will improve GDBM's "crash tolerance"), I believe the following links give a detailed answer: https://queue.acm.org/detail.cfm?id=3487353 and https://www.gnu.org.ua/software/gdbm/manual/Crash-Tolerance.html.

jkeenan commented:
> Hi, Jim
> 
> I was able to build the stuff on FreeBSD 12.0 (and 13.0, for what it's worth) without any problem. In your build case, I don't know what regular_configure is, so I can only suppose that there's something wrong with the configuration options it uses. If you can give me some insight into what it does, perhaps I could investigate it further.
> 

```
sh ./Configure -des -Dusedevel \
            -Duseithreads \
            -Doptimize="-O2 -pipe -fstack-protector -fno-strict-aliasing"
```
This invocation was chosen so as to be close to what the vendor `perl` is like on FreeBSD.  But I doubt this is the source of the problem.  I use it in our smoke-testing on FreeBSD several times a week, so it's "road-tested."

> Re "recent years maintaining GDBM_File has been a PITA", my impression is that too little effort has been invested in maintaining it. That's why I started to do it myself. As a maintainer of GNU dbm I'm interested in that GDBM_File adequately reflects the gdbm functionality and provides the same possibilities.
> 

Here's an example of the problems which I face when a new Issue is filed that pertains to GDBM_File:
```
# "Vendor" perl
$ perl -v | head -2 | tail -1
This is perl 5, version 32, subversion 1 (v5.32.1) built for amd64-freebsd-thread-multi

$ perl -MGDBM_File -e 'my $str = GDBM_File->GDBM_version; print qq{$str\n};'
String found where operator expected at /usr/local/lib/perl5/site_perl/mach/5.32/GDBM_File.pm line 74, near "XSLoader::load 'GDBM_File'"
	(Do you need to predeclare XSLoader::load?)
syntax error at /usr/local/lib/perl5/site_perl/mach/5.32/GDBM_File.pm line 74, near "XSLoader::load 'GDBM_File'"
Compilation failed in require.
BEGIN failed--compilation aborted.

# perl-5.34.0
$ ./bin/perl -v | head -2 | tail -1
This is perl 5, version 34, subversion 0 (v5.34.0) built for amd64-freebsd-thread-multi
[perlmonger: 5.34.0] $ ./bin/perl -Ilib -MGDBM_File -e 'my $str = GDBM_File->GDBM_version; print qq{$str\n};'
1.18.1

# blead as of today
$ git describe
v5.35.4-83-gef3c47f87b
$ ./perl -Ilib -MGDBM_File -e 'my $str = GDBM_File->GDBM_version; print qq{$str\n};'
1.19
```
I don't claim that these calls on the command-line are entirely relevant to my build failure.  But they reflect why I find GDBM_File such a puzzlement.  See also the 4 open, GDBM_File-related tickets [here](https://github.com/Perl/perl5/issues?q=is%3Aissue+sort%3Aupdated-desc+label%3Aext%2FGDBM_File+is%3Aopen).

> As to the last question (why this will improve GDBM's "crash tolerance"), I believe the following links give a detailed answer: https://queue.acm.org/detail.cfm?id=3487353 and https://www.gnu.org.ua/software/gdbm/manual/Crash-Tolerance.html.

I suspect that, once we get closer to merging this p.r. into blead, a one-paragraph summary of those pages should go into the commit message.

graygnuorg commented:
> See also the 4 open, GDBM_File-related tickets here.

The [#19149] is fixed in one of my prior commits, as tonycoz observed.
In [#12957] the reporter is trying to do something that cannot be done using a hashed database, and the documentation advises against it. (I commented on the issue with the details).

Unfortunately I cannot check the remaining two because I don't have access to the OS in question.

jkeenan commented:
I have duped this branch into new branch `smoke-me/jkeenan/gray/gdbm-crashtolerance` in order to make this available for smoke-testing on other platforms.

jkeenan commented:
The build is also failing on one of our CI rigs: macOS (catalina) xcode 12 (-Duseithreads -Duseshrplib).  See https://github.com/Perl/perl5/pull/19182/checks?check_run_id=3848352145.
```
2021-10-09T17:16:12.7200810Z cc -c   -fno-common -DPERL_DARWIN -mmacosx-version-min=10.15 -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -DPERL_USE_SAFE_PUTENV -Wall -Werror=pointer-arith -Werror=vla -Wextra -Wc++-compat -Wwrite-strings -Werror=declaration-after-statement -Wno-error=implicit-function-declaration -O3   -DVERSION=\"1.21\" -DXS_VERSION=\"1.21\"  "-I../.."   GDBM_File.c
2021-10-09T17:16:13.0035060Z GDBM_File.xs:227:13: error: use of undeclared identifier 'my_perl'
2021-10-09T17:16:13.0134830Z         t = SvUV(val);
2021-10-09T17:16:13.0236610Z             ^
2021-10-09T17:16:13.0338910Z ../../sv.h:1847:47: note: expanded from macro 'SvUV'
2021-10-09T17:16:13.0440620Z #define SvUV(sv) (SvUOK_nog(sv) ? SvUVX(sv) : sv_2uv(sv))
2021-10-09T17:16:13.0542110Z                                               ^
2021-10-09T17:16:13.0646090Z ../../sv.h:2155:20: note: expanded from macro 'sv_2uv'
2021-10-09T17:16:13.0687680Z #define sv_2uv(sv) sv_2uv_flags(sv, SV_GMAGIC)
2021-10-09T17:16:13.0689250Z                    ^
2021-10-09T17:16:13.0690950Z ../../embed.h:583:45: note: expanded from macro 'sv_2uv_flags'
2021-10-09T17:16:13.0691890Z #define sv_2uv_flags(a,b)       Perl_sv_2uv_flags(aTHX_ a,b)
2021-10-09T17:16:13.0696060Z                                                   ^
2021-10-09T17:16:13.0697400Z ../../perl.h:161:18: note: expanded from macro 'aTHX_'
2021-10-09T17:16:13.0698210Z #  define aTHX_         aTHX,
2021-10-09T17:16:13.0698810Z                         ^
2021-10-09T17:16:13.0699880Z ../../perl.h:156:16: note: expanded from macro 'aTHX'
2021-10-09T17:16:13.0700710Z #  define aTHX  my_perl
2021-10-09T17:16:13.0703950Z                 ^
2021-10-09T17:16:13.0705350Z GDBM_File.xs:229:19: error: use of undeclared identifier 'my_perl'
2021-10-09T17:16:13.0706730Z         char *s = SvPV_nolen(val);
2021-10-09T17:16:13.0707380Z                   ^
2021-10-09T17:16:13.0708710Z ../../sv.h:1898:20: note: expanded from macro 'SvPV_nolen'
2021-10-09T17:16:13.0711910Z      ? SvPVX(sv) : sv_2pv_flags(sv, 0, SV_GMAGIC))
2021-10-09T17:16:13.0712630Z                    ^
2021-10-09T17:16:13.0713890Z ../../embed.h:580:47: note: expanded from macro 'sv_2pv_flags'
2021-10-09T17:16:13.0715290Z #define sv_2pv_flags(a,b,c)     Perl_sv_2pv_flags(aTHX_ a,b,c)
2021-10-09T17:16:13.0716210Z                                                   ^
2021-10-09T17:16:13.0717870Z ../../perl.h:161:18: note: expanded from macro 'aTHX_'
2021-10-09T17:16:13.0718820Z #  define aTHX_         aTHX,
2021-10-09T17:16:13.0720460Z                         ^
2021-10-09T17:16:13.0725250Z ../../perl.h:156:16: note: expanded from macro 'aTHX'
2021-10-09T17:16:13.0980900Z #  define aTHX  my_perl
2021-10-09T17:16:13.0981280Z                 ^
2021-10-09T17:16:13.0982660Z GDBM_File.c:1471:6: warning: unused variable 'RETVAL' [-Wunused-variable]
2021-10-09T17:16:13.0983230Z         int     RETVAL;
2021-10-09T17:16:13.0983540Z                 ^
2021-10-09T17:16:13.0984450Z GDBM_File.c:1472:2: warning: unused variable 'targ' [-Wunused-variable]
2021-10-09T17:16:13.0985010Z         dXSTARG;
2021-10-09T17:16:13.0985300Z         ^
2021-10-09T17:16:13.0986080Z ../../XSUB.h:189:28: note: expanded from macro 'dXSTARG'
2021-10-09T17:16:13.0987120Z #define dXSTARG SV * const targ = ((PL_op->op_private & OPpENTERSUB_HASTARG) \
2021-10-09T17:16:13.0987650Z                            ^
2021-10-09T17:16:13.0988050Z 2 warnings and 2 errors generated.
2021-10-09T17:16:13.0988490Z make[1]: *** [GDBM_File.o] Error 1
2021-10-09T17:16:13.0989030Z Unsuccessful make(ext/GDBM_File): code=512 at make_ext.pl line 584.
2021-10-09T17:16:13.0989640Z make: *** [lib/auto/GDBM_File/GDBM_File.bundle] Error 25
2021-10-09T17:16:13.0990180Z make: *** Waiting for unfinished jobs....
2021-10-09T17:16:13.2408670Z rm -f ../../../lib/auto/Encode/Unicode/Unicode.bundle
2021-10-09T17:16:13.2440300Z cc  -mmacosx-version-min=10.15 -bundle -undefined dynamic_lookup -L/usr/local/lib -fstack-protector-strong  Unicode.o  -o ../../../lib/auto/Encode/Unicode/Unicode.bundle  \
2021-10-09T17:16:13.2441870Z 	      \
2021-10-09T17:16:13.2442150Z 	  
2021-10-09T17:16:13.3007530Z chmod 755 ../../../lib/auto/Encode/Unicode/Unicode.bundle
2021-10-09T17:16:13.3321510Z ##[error]Process completed with exit code 2.
```

graygnuorg committed:
Bugfix

* ext/GDBM_File/GDBM_File.xs (is_true): Add missing context parameter.
All uses changed
(gdbm_dump): Change return type to void.

graygnuorg commented:
Thanks for additional details.  I was able to reproduce this using -Dusethreads -Duseithreads. The choice of compiler and OS turned out to be irrelevant. I have pushed commit a9ce22f7 that fixes the bug.

jkeenan commented:
> Thanks for additional details. I was able to reproduce this using -Dusethreads -Duseithreads. The choice of compiler and OS turned out to be irrelevant. I have pushed commit [a9ce22f](https://github.com/Perl/perl5/commit/a9ce22f746d454add2aa0c19fe25e1dc89bdf5ba) that fixes the bug.

Looks good.  I have cherry-picked your last commit onto the smoke-me/jkeenan/gray/gray/gdbm-crashtolerance branch so as to get another round of smoke-testing on other platforms.  (Other FreeBSD rigs were showing failures like mine; see http://perl.develop-help.com/?b=smoke-me%2Fjkeenan%2Fgray%2Fgdbm-crashtolerance).

I would like to get this further reviewed by @tonycoz.

tonycoz commented on ext/GDBM_File/GDBM_File.xs:
```diff
@@ -624,6 +650,116 @@ gdbm_count(db)
+                kw = SvPV_nolen(sv);
+
+                if (strcmp(kw, "restore_mode") == 0) {
+                    if (!is_true(val))
```
Why don't you use SvTRUE() here? (and many other places)

tonycoz commented on ext/GDBM_File/GDBM_File.xs:
```diff
@@ -624,6 +650,116 @@ gdbm_count(db)
+                if (strcmp(kw, "mode") == 0) {
+                    if (SvIOK(val)) {
+                        mode = SvUV(val) & 0777;
+                    }
```
In general, it is better to not check flags directly.

In both cases here if the key or value has any magic the magic will be ignored for the flag (check).  It also behaves badly with overrides.

If you want to check whether the key is defined you might do:
```
  SvGETMAGIC(sv);
  if (!SvOK(sv))
    croak(...);
```

but it is simpler to just call SvPV_nolen() without the flag check, which will call magic itself, and warn if the user has warnings enabled for an undefined value, and handle overloads correctly.

Similarly for SvIOK(), if `val` has any magic the flags might be stale, but IOK might not be set if `val` is an overloaded value, which is fairly likely if the code is in the context of `use bigint;`.

tonycoz commented on ext/GDBM_File/GDBM_File.xs:
```diff
@@ -624,6 +650,116 @@ gdbm_count(db)
+                SV *sv = ST(i);
+                SV *val = ST(i+1);
+
+                if (!SvPOK(sv))
```
Similar to my above comments on checking flags.

graygnuorg committed:
Use SvTRUE to check for true value.  Remove unnecessary checks for SV types.
*  ext/GDBM_File/GDBM_File.xs

jkeenan commented:
> Hi Tony,
> 
> Thanks for suggestions. I've pushed a patch.

I have updated the `smoke-me/jkeenan/gray/gdbm-crashtolerance` branch with this patch.

graygnuorg committed:
Remove remaining type checks

* ext/GDBM_File/GDBM_File.xs: Remove remaining uses of Sv[IPU]OK
* ext/GDBM_File/t/snapshot.t: Remove formfeeds.

graygnuorg commented:
> There's still some flag checks, SvPOK() in gdbm_dump(), gdbm_load().

I removed them.  I left SvROK in gdbm_recover, but inserted a SvGETMAGIC before them.

> snapshot.t includes what looks like form feed characters on some lines

I use form feeds  to delimit portions of text.  Didn't know they are not welcome.  Removed them as well.

tonycoz commented on ext/GDBM_File/GDBM_File.xs:
```diff
@@ -878,11 +871,7 @@ gdbm_flags(db, ...)
                 opcode = GDBM_GETMAXMAPSIZE;
             } else {                                      
                 opcode = GDBM_SETMAXMAPSIZE;
-                sv = ST(1);                               
-                if (!SvUOK(sv)) {                         
-                    croak("%s: bad argument type", opt_names[ix]);           
-                }                                         
-                c_uv = SvUV(sv);                          
+                c_uv = SvUV(ST(1));
```
This eliminates the last use of sv, but leaves the declaration.

GDBM_File.c: In function ‘XS_GDBM_File_flags’:
GDBM_File.xs:820:13: warning: unused variable ‘sv’ [-Wunused-variable]
         SV *sv;
             ^~

graygnuorg committed:
Remove unused variable

xenu committed:
install libgdbm and libdb in GitHub Actions

This ensures that DB_File and GDBM_File will be built and tested.

xenu commented:
Apparently, our GitHub actions don't install libdb and libgdbm, so DB_File and GDBM_File don't get built. I've pushed a commit to your branch which should rectify that problem. As for op/alarm.t, don't worry about it. Some of our tests are flaky on Cygwin.

xenu commented:
It seems my change has uncovered a problem: GDBM_File randomly causes fatal errors in LeakSanitizer:
```
2021-10-13T16:38:42.9895068Z Tracer caught signal 11: addr=0x0 pc=0x4d6544 sp=0x7fc328bf7d40
2021-10-13T16:38:42.9895940Z ==47346==LeakSanitizer has encountered a fatal error.
2021-10-13T16:38:42.9898747Z ==47346==HINT: For debugging, try setting environment variable LSAN_OPTIONS=verbosity=1:log_threads=1
2021-10-13T16:38:42.9899882Z ==47346==HINT: LeakSanitizer does not work under ptrace (strace, gdb, etc)
2021-10-13T16:38:42.9938747Z ../cpan/Memoize/t/errors.t ........................................... 
```

It happens indeterministically in various tests that call GDBM_File, not just the one I pasted above. It **isn't** caused by the changes in this branch, I managed to reproduce it by cherry-picking my github actions config fix to a blead-based branch.

I'm not sure if it's a bug in ASan, GDBM_File or both, but either way, it should be resolved in a separate ticket.

I've merged and squashed this branch (minus my github actions changes) as 1d7b70436255bf0e7e5c9a8db29ffbf04371b978.
